tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1439 @@
+    auto* self = reinterpret_cast<ListTask*>(task);
+    execution::set_starting(self->receiver_, [self] {
+      self->cancelled_.store(true, std::memory_order_relaxed);
+    });
+    self->owner_->executor()(
+        [state = IntrusivePtr<ListTask>(self, internal::adopt_object_ref)] {
+          state->IssueRequest();
+        });
+  }
+
+  void Retry() { IssueRequest(); }
+
+  void IssueRequest() {
+    if (is_cancelled()) {
+      execution::set_done(receiver_);
+      execution::set_stopping(receiver_);
+      return;
+    }
+
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
+    auto request_builder = S3RequestBuilder("GET", resource_)
+                            .AddQueryParameter("list-type", "2");
+
+
+    if (auto& prefix = options_.range.inclusive_min; !prefix.empty()) {
+      if (options_.strip_prefix_length) {
+         prefix = prefix.substr(0, options_.strip_prefix_length);
+      }
+      request_builder.AddQueryParameter("prefix", prefix);
+    }
+
+    if (!continuation_token_.empty()) {
+      request_builder.AddQueryParameter("continuation-token", continuation_token_);
+    }
+
+    auto maybe_credentials = owner_->GetCredentials();
+    if(!maybe_credentials.ok()) {
+      execution::set_error(receiver_, std::move(maybe_credentials).status());
+      execution::set_stopping(receiver_);
+      return;
+    }
+
+    S3Credentials credentials;
+
+    if(maybe_credentials.value().has_value()) {
+      credentials = std::move(*maybe_credentials.value());
+    }
+
+    start_time_ = absl::Now();
+
+    auto request = request_builder
+        .AddHeader(absl::StrCat("host: ", owner_->host_))
+        .AddHeader(absl::StrCat("x-amz-content-sha256: ", kEmptySha256))
+        .AddHeader(absl::FormatTime("x-amz-date: %Y%m%dT%H%M%SZ", start_time_, absl::UTCTimeZone()))
+        .BuildRequest(
+          credentials.GetAccessKey(),
+          credentials.GetSecretKey(),
+          owner_->aws_region_,
+          kEmptySha256,
+          start_time_);
+
+
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_REQUESTS)
+        << "List: " << request;
+
+    auto future = owner_->transport_->IssueRequest(request, {});
+    future.ExecuteWhenReady(WithExecutor(
+        owner_->executor(), [self = IntrusivePtr<ListTask>(this)](
+                                ReadyFuture<HttpResponse> response) {
+          self->OnResponse(response.result());
+        }));
+  }
+
+  void OnResponse(const Result<HttpResponse>& response) {
+    auto status = OnResponseImpl(response);
+    // OkStatus are handled by OnResponseImpl
+    if (absl::IsCancelled(status)) {
+      execution::set_done(receiver_);
+      execution::set_stopping(receiver_);
+      return;
+    }
+    if (!status.ok()) {
+      execution::set_error(receiver_, std::move(status));
+      execution::set_stopping(receiver_);
+      return;
+    }
+  }
+
+  /// @brief Find the starting position of the tag or the position immediately after the tag
```

Comment:
These seem like helpers that might fit well with other utility methods such as validate.h?

Reply:
Agreed. I've moved these functions into `s3_metadata.{h,cc}`, and added a test case.


```
/// @brief Get tag contents within the supplied XML payload
///
/// This should primarily be used for obtaining the contents of inner tags.
/// Use FindTag to set the initial search position by traversing XML outer tags.
///
/// @param data XML payload
/// @param open_tag Opening tag
/// @param close_tag Closing tag
/// @param pos Initial searching position. Updated with position immediately
///            after the closing tag.
/// @return Tag contents
Result<std::string_view> GetTag(std::string_view data, std::string_view open_tag, std::string_view close_tag,
                                                    std::size_t * pos);
```

For the record, I dislike the update of `pos` in `GetTag`. I would prefer a signature like
`Result<std::pair<std::string_view, std::size_t>> GetTag(data, open_tag, close_tag)`, but the combination of tuple destructuring and macros in `TENSORSTORE_ASSIGN_OR_RETURN(auto [tag, newpos], GetTag(...))` doesn't seem to work. Any tips here?

Reply:
> For the record, I dislike the update of pos in GetTag. I would prefer a signature like
Result<std::pair<std::string_view, std::size_t>> GetTag(data, open_tag, close_tag), but the combination of tuple destructuring and macros in TENSORSTORE_ASSIGN_OR_RETURN(auto [tag, newpos], GetTag(...)) doesn't seem to work. Any tips here?

Maybe return a struct

Reply:
I went with a `TagAndPosition` struct.